### PR TITLE
Fix scatter size check

### LINKED ISSUE #1653: Tensor.scatter_ accepts bogus input where index size > src size
e.g.:
```
>>> torch.zeros(3,4).scatter_(0,torch.LongTensor([[0,0,0,0],[1,1,1,1],[2,2,2,2]]),torch.arange(1,9).resize_(2,4))

 1.0000e+00  2.0000e+00  3.0000e+00  4.0000e+00
 5.0000e+00  6.0000e+00  7.0000e+00  8.0000e+00
 2.7716e+20  7.2128e+22  4.4653e+30  7.2708e+31
[torch.FloatTensor of size 3x4]
```

valgrind reports reading off the end of the tensor:
```
==2959389== Invalid read of size 4
==2959389==    at 0x210EE636: THFloatTensor_scatter (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x169365A8: THPFloatTensor_scatter_(_object*, _object*, _object*) (TensorMethods.cpp:3493)
==2959389==    by 0x4F00901: _PyCFunction_FastCallDict (methodobject.c:231)
==2959389==    by 0x4F85F4B: call_function (ceval.c:4788)
==2959389==    by 0x4F88BBC: _PyEval_EvalFrameDefault (ceval.c:3275)
==2959389==    by 0x4F844BF: _PyEval_EvalCodeWithName (ceval.c:4119)
==2959389==    by 0x4F84942: PyEval_EvalCodeEx (ceval.c:4140)
==2959389==    by 0x4F8498A: PyEval_EvalCode (ceval.c:695)
==2959389==    by 0x4FB9345: run_mod (pythonrun.c:980)
==2959389==    by 0x4FB9345: PyRun_InteractiveOneObject (pythonrun.c:233)
==2959389==    by 0x4FB96AD: PyRun_InteractiveLoopFlags (pythonrun.c:112)
==2959389==    by 0x4FB97EB: PyRun_AnyFileExFlags (pythonrun.c:74)
==2959389==    by 0x4FD3BA8: run_file (main.c:320)
==2959389==    by 0x4FD3BA8: Py_Main (main.c:781)
==2959389==  Address 0x12f19840 is 0 bytes after a block of size 32 alloc'd
==2959389==    at 0x4C26B0F: malloc (in /usr/local/fbcode/gcc-5-glibc-2.23/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==2959389==    by 0x2106CC79: THAlloc (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x2106E037: THFloatStorage_resize (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x21087194: THFloatTensor_resize4d (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x2110DEF6: THFloatTensor_range (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x16981F53: THPFloatTensor_stateless_arange(_object*, _object*, _object*) (TensorMethods.cpp:3428)
==2959389==    by 0x4F00CD8: PyCFunction_Call (methodobject.c:98)
==2959389==    by 0x4EA9CE5: PyObject_Call (abstract.c:2246)
==2959389==    by 0x167EEAB1: THPUtils_dispatchStateless(_object*, char const*, _object*, _object*) (utils.cpp:154)
==2959389==    by 0x167C221F: THPModule_arange(_object*, _object*, _object*) (Module.cpp:300)
==2959389==    by 0x4F00901: _PyCFunction_FastCallDict (methodobject.c:231)
==2959389==    by 0x4F85F4B: call_function (ceval.c:4788)
==2959389==
```

### PR DESCRIPTION
**Pull request by ssnl**
Fixes #1653 

This ensures the following condition on `scatter` for both cpu and cuda:

1. index.size[d] <= output.size[d] for d != dim 
2. index.size[d] <= src.size[d] for all d

The drawback is introducing a weird macro argument. Let me know if I should think of other ways,

### CODE REVIEW

File: torch/lib/THC/generic/THCTensorScatterGather.cu [hunk 1]
```diff
@@ -111,16 +111,15 @@ void THCTensor_(scatter)(THCState* state, THCTensor *tensor, int dim, THCudaLong
...
-             "Index tensor must have the same size as input tensor.");
-  THLongStorage_free(indexDims);
 
   for (int d = 0; d < THCTensor_(nDimension)(state, tensor); d++) {
+    int64_t indexSizeD = THCudaLongTensor_size(state, index, d);
     if (d != dim) {
-      THArgCheck(THCTensor_(size)(state, tensor, d) == THCTensor_(size)(state, src, d), 4,
-                 "Input tensor must have same size as output tensor apart from the specified dimension");
+      THArgCheck(indexSizeD <= THCTensor_(size)(state, tensor, d), 3,
+                 "Index tensor must not have larger size than output tensor apart from the specified dimension");
```
Comment (soumith) [inline torch/lib/THC/generic/THCTensorScatterGather.cu]: while you're touching this code, can you fix the error message to also report the sizes of the tensors (like the CPU error)

Comment (SsnL) [reply torch/lib/THC/generic/THCTensorScatterGather.cu]: On it.

**Comment by SsnL** [comment]
Updated cuda side error msg with size information.

File: torch/lib/TH/THTensorDimApply.h [hunk 2]
```diff
@@ -1,7 +1,32 @@
 #ifndef TH_TENSOR_DIM_APPLY_INC
 #define TH_TENSOR_DIM_APPLY_INC
 
-#define TH_TENSOR_DIM_APPLY3(TYPE1, TENSOR1, TYPE2, TENSOR2, TYPE3, TENSOR3, DIMENSION, CODE) \
+#define TH_TENSOR_DIM_APPLY3_SIZE_EQ_EXCEPT_DIM(TENSOR1, TENSOR2, TENSOR3, DIMENSION) \
```
Comment (apaszke) [inline torch/lib/TH/THTensorDimApply.h]: Can you add a comment saying how this macro differs from all other applies? We have so many variants that I don't even know what they're for at this point.

**Review by apaszke — COMMENTED** [review]
Didn't finish reviewing. Just asking for comments

File: torch/lib/TH/THTensorDimApply.h (hunk 2 above)
Comment (SsnL) [reply torch/lib/TH/THTensorDimApply.h]: Sure, I will add one soon. 

As a quick explanation, this is a part of original `TH_TENSOR_DIM_APPLY3`, extracted out meant to refactor the size check so that scatter can apply its own check. The `TH_TENSOR_DIM_APPLY3` is still there below.

File: torch/lib/THC/generic/THCTensorScatterGather.cu [hunk 3]
```diff
@@ -111,16 +111,17 @@ void THCTensor_(scatter)(THCState* state, THCTensor *tensor, int dim, THCudaLong
...
-  THLongStorage_free(indexDims);
 
   for (int d = 0; d < THCTensor_(nDimension)(state, tensor); d++) {
+    int64_t indexSizeD = THCudaLongTensor_size(state, index, d);
     if (d != dim) {
-      THArgCheck(THCTensor_(size)(state, tensor, d) == THCTensor_(size)(state, src, d), 4,
-                 "Input tensor must have same size as output tensor apart from the specified dimension");
+      THArgCheck(indexSizeD <= THCTensor_(size)(state, tensor, d), 3,
+                 "Index tensor must not have larger size than output tensor apart from the specified dimension %d, but got index %s output %s",
+                 dim, THCudaLongTensor_sizeDesc(state, index).str, THCTensor_(sizeDesc)(state, tensor).str);
```
Comment (apaszke) [inline torch/lib/THC/generic/THCTensorScatterGather.cu]: Shouldn't the output tensor have a matching size with index? We should just resize it

File: torch/lib/THC/generic/THCTensorScatterGather.cu [hunk 4]
```diff
@@ -111,16 +111,17 @@ void THCTensor_(scatter)(THCState* state, THCTensor *tensor, int dim, THCudaLong
...
     }
+    THArgCheck(indexSizeD <= THCTensor_(size)(state, src, d), 3,
```
Comment (apaszke) [inline torch/lib/THC/generic/THCTensorScatterGather.cu]: Are you sure this is really true? I think it should be fine to e.g. repeat a tensor using scatter

**Review by apaszke — COMMENTED** [review]
Either I don't understand something, or it's not a complete fix for the issue linked. Some size checks seem to be unnecessarily strict, and the main issue mentioned there is that we accept 2 as index for the 0th dim, even if it's out of bound. This requires us to inspect the index data, which is not done in this PR.

File: torch/lib/THC/generic/THCTensorScatterGather.cu (hunk 4 above)
Comment (apaszke) [reply torch/lib/THC/generic/THCTensorScatterGather.cu]: I can't seem to find this assertion before, so I'd assume that the kernel is ok with repeating some elements

**Comment by SsnL** [comment]
According to my understanding of `scatter`, what it does is:
for specified `dim`
 for all `idx = index[x_1, x_2, ..., x_dim, ..., x_n]`, 
    let `output[x_1, x_2, ..., idx, ..., x_n] = src[x_1, x_2, ..., x_dim, ..., x_n]`

The issue linked is talking about the size relation between `index` and `src`, i.e. `index` has larger size than `src` and thus scatters uninitialized memory data to `output`.

I wonder how scatter can be used to repeat elements. If I understand correctly, it can only copy each value to a certain position.

File: torch/lib/THC/generic/THCTensorScatterGather.cu (hunk 4 above)
Comment (soumith) [reply torch/lib/THC/generic/THCTensorScatterGather.cu]: hmm yea Adam makes a good point here, maybe we can remove this assertion. (sorry i didn't think about this case before)

Comment (soumith) [reply torch/lib/THC/generic/THCTensorScatterGather.cu]: actually ignore my comment. Simon pointed out that `index` is actually for the output tensor, so repeating a tensor using scatter doesn't hold (it doesn't work in an intuitive way):
http://pytorch.org/docs/master/tensors.html?highlight=scatter#torch.Tensor.scatter_

**Comment by apaszke** [comment]
Yeah I might have confused some parts with `gather`. Still, do we check for out of bound indices in index tensor?

**Comment by SsnL** [comment]
I just tested. Both CPU and CUDA version will throw error in that case. :)
